receiver/m365receiver/client.go

```diff
@@ -0,0 +1,424 @@
+func (m *m365Client) GetToken() error {
+	formData := url.Values{
+		"grant_type":    {"client_credentials"},
+		"scope":         {m.scope},
+		"client_id":     {m.clientID},
+		"client_secret": {m.clientSecret},
+	}
+
+	requestBody := strings.NewReader(formData.Encode())
+
+	req, err := http.NewRequest("POST", m.authEndpoint, requestBody)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot resp err if present
+	if resp.StatusCode != 200 {
+		// get error code
+		respErr := struct {
+			Err string `json:"error"`
+		}{}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(body, &respErr)
+		if err != nil {
+			return err
+		}
+		//match on error code
+		switch respErr.Err {
+		case "unauthorized_client":
+			return fmt.Errorf("the provided client_id is incorrect or does not exist within the given tenant directory")
+		case "invalid_client":
+			return fmt.Errorf("the provided client_secret is incorrect or does not belong to the given client_id")
+		case "invalid_request":
+			return fmt.Errorf("the provided tenant_id is incorrect or does not exist")
+		}
+		return fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	// read in access token
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	token := struct {
+		Token string `json:"access_token"`
+	}{}
+	err = json.Unmarshal(body, &token)
+	if err != nil {
+		return err
+	}
+	m.token = token.Token
+	return nil
+}
+
+// function for getting metrics data
+func (m *m365Client) GetCSV(endpoint string) ([]string, error) {
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return []string{}, err
+	}
+
+	req.Header.Set("Authorization", m.token)
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return []string{}, err
+	}
+
+	//troubleshoot resp err if present
+	if resp.StatusCode != 200 {
+		//get error code
+		respErr, err := m.readInErr(resp)
+		if err != nil {
+			return []string{}, err
+		}
+		if respErr.Error.Code == "InvalidAuthenticationToken" {
+			return []string{}, fmt.Errorf("access token invalid")
+		}
+		return []string{}, fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	csvReader := csv.NewReader(resp.Body)
+
+	//parse out 2nd line & return csv data
+	_, err = csvReader.Read()
+	if err != nil {
+		return []string{}, err
+	}
+	data, err := csvReader.Read()
+	if err != nil {
+		if err == io.EOF { //no data in report, scraper should still run
+			return []string{}, nil
+		}
+		return []string{}, err
+	}
+
+	return data, nil
+}
+
+// function for getting log data
+func (m *m365Client) GetJSON(ctx context.Context, endpoint string, end string, start string) (logData, error) {
```

Comment:
Just noticed that only this function accepts context as a parameter, they probably all should & request with context.